src/slice.cpp

```
@@ -83,7 +83,7 @@ DataFrame slice_grouped(GroupedDataFrame gdf, const QuosureList& dots) {
       // positive indexing
       int ntest = g_test.size();
       for (int j = 0; j < ntest; j++) {
-        if (!(g_test[j] > nr || g_test[j] == NA_INTEGER)) {
+        if (!(g_test[j] > nr || g_test[j] < 1)) {
```

We currently fail if positive and negative indexes are used, but this change seems to silently ignore negatives (including `NA`):

``` r
dplyr::slice(iris, c(1, -1))
#> Error in slice_impl(.data, dots): Indices must be either all positive or all negative, not a mix of both. Found 1 positive indices and 1 negative indices
```

> Created on 2018-03-12 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

Also, I'm struggling with `!(... || ...)` , can you please refactor?

---

Do we have a test for `slice(c(-1, 1))` ?

---

We do, but perhaps not for `slice(c(1, -1))` ?

---

When we are in the `positive indexing` branch we already know that we either have only positive

---

I see. Can you please add comments to the fact that we're ignoring both 0 and NA here (because NA happens to be -2³¹)?

I'll work on reducing code duplication in #2311.